pkg/p2p/client.go

```diff
@@ -0,0 +1,446 @@
+	// The size of the sending channel used to buffer
+	// messages before they go to gRPC.
+	SendChannelSize int
+	// The maximum duration for which messages wait to be batched.
+	BatchSendInterval time.Duration
+	// The maximum size in bytes of a batch.
+	MaxBatchBytes int
+	// The maximum number of messages in a batch.
+	MaxBatchCount int
+	// The limit of the rate at which the connection to the server is retried.
+	RetryRateLimitPerSecond float64
+	// The dial timeout for the gRPC client
+	DialTimeout time.Duration
+	// The advertised address of this node. Used for logging and monitoring purposes.
+	AdvertisedAddr string
+	// The version of the client for compatibility check.
+	// It should be in semver format. Empty string means no check.
+	ClientVersion string
+}
+
+// MessageClient is a client used to send peer messages.
+// `Run` must be running before sending any message.
+type MessageClient struct {
+	sendCh chan *p2p.MessageEntry
+
+	topicMu sync.RWMutex
+	topics  map[string]*topicEntry
+
+	senderID NodeID
+
+	closeCh  chan struct{}
+	isClosed atomic.Bool
+
+	// config is read only
+	config *MessageClientConfig
+
+	// newSender is used to create a new sender.
+	// It can be replaced to unit test MessageClient.
+	newSenderFn func(clientStream) clientBatchSender
+
+	connector clientConnector
+}
+
+type topicEntry struct {
+	sentMessageMu sync.Mutex
+	sentMessages  deque.Deque
+
+	nextSeq  atomic.Int64
+	ack      atomic.Int64
+	lastSent atomic.Int64
+}
+
+// NewMessageClient creates a new MessageClient
+// senderID is an identifier for the local node.
+func NewMessageClient(senderID NodeID, config *MessageClientConfig) *MessageClient {
+	return &MessageClient{
+		sendCh:   make(chan *p2p.MessageEntry, config.SendChannelSize),
+		topics:   make(map[string]*topicEntry),
+		senderID: senderID,
+		closeCh:  make(chan struct{}),
+		config:   config,
+		newSenderFn: func(stream clientStream) clientBatchSender {
+			return newClientBatchSender(stream, config.MaxBatchBytes, config.MaxBatchCount)
+		},
+		connector: newClientConnector(),
+	}
+}
+
+// Run launches background goroutines for MessageClient to work.
+func (c *MessageClient) Run(
+	ctx context.Context, network, addr string,
+	receiverID NodeID,
+	credential *security.Credential,
+) (ret error) {
+
+	defer func() {
+		c.isClosed.Store(true)
+		close(c.closeCh)
+
+		log.Info("peer message client exited",
+			zap.String("addr", addr),
+			zap.String("capture-id", receiverID),
+			zap.Error(ret))
+	}()
+
+	metricsClientCount := clientCount.With(prometheus.Labels{
+		"to": addr,
+	})
+	metricsClientCount.Inc()
+	defer metricsClientCount.Dec()
+
+	rl := rate.NewLimiter(rate.Limit(c.config.RetryRateLimitPerSecond), 1)
+	epoch := int64(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		default:
+		}
+
+		if err := rl.Wait(ctx); err != nil {
+			return errors.Trace(err)
+		}
+
+		gRPCClient, release, err := c.connector.Connect(clientConnectOptions{
```

Comment:
Should we move this `Connect` out of the loop ?

Reply:
I prefer reconnecting (i.e., redialing) the server in each iteration of the retry loop because more errors can be resolved by redialing than by simply resending a streaming request.